007-extension-flag.md

```diff
@@ -0,0 +1,43 @@
+| Title  | Implement --extension flag  |
+|--------|-----------------------------|
+| Author | @bmeck                      |
+| Status | DRAFT                       |
+| Date   | 2017-08-09T10:00:00-05:00   |
+
+## Description
+
+The Node CLI currently accepts input in both filename (`node app.js`) and opaque text stream form (`node -e '123'` and `node <app.js`). This behavior always results in running the stream as a CommonJS goal. With the upcoming usage of ESM and WASM there will be cases where developers wish to pipe other forms of text to the CLI.
+
+This proposal is to implement an `--entry-extension` CLI flag for the cases which may want to use a stream that does not have a file extension.
+
+## Example
+
+```sh
+node --entry-extension=js <app.js
+node --entry-extension=js -e $(cat <app.js)
+node --entry-extension=wasm <app.wasm
+node --entry-extension=wasm -e $(cat <app.wasm)
+node --entry-extension=mjs <app.mjs
+node --entry-extension=mjs -e $(cat <app.mjs)
+```
+
+## Considerations
+
+If mixing `--entry-extension` with a non-stream argument such as a file path or entry URL, Node should exit with an error prior to any evaluation. This also means that `--entry-extension` cannot be used to override the extension of file paths.
+
+```sh
+# exits with an error
+node --entry-extension=json app.js
```

Comment:
Might be a good example to add so that people don't think it's the mismatch that errors